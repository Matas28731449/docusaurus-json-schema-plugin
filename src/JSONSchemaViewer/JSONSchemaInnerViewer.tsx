import React from "react"

import type { JSONSchema7 } from "json-schema"

type InnerViewerProperties = {
  schema: JSONSchema7
}

function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  return <></>
}

export default JSONSchemaInnerViewer
