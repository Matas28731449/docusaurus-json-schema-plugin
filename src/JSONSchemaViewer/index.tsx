import React, { useEffect, useState } from "react"
import Layout from "@theme/Layout"
import Translate from "@docusaurus/Translate"
import AllOfMerger from "json-schema-merge-allof"
import { Resolver } from "@stoplight/json-ref-resolver"

import type { JSONSchema7 } from "json-schema"

import JSONSchemaViewer from "./JSONSchemaInnerViewer"

type ViewerProperties = {
  loadSchema: () => Promise<JSONSchema7>
}

// merged representation of allOf array of schemas
async function mergeAllOf(
  schema: JSONSchema7
): Promise<Omit<JSONSchema7, "allOf">> {
  const mergedSchema = AllOfMerger(schema, {
    // Technically valid, but only interested on fields with name, at least for now ...
    ignoreAdditionalProperties: true,
  })

  return Promise.resolve(mergedSchema as Omit<JSONSchema7, "allOf">)
}

function JSONSchemaEditor(props: ViewerProperties): JSX.Element {
  const [schema, setSchema] = useState(
    undefined as undefined | Omit<JSONSchema7, "allOf">
  )
  const [fetchError, setFetchError] = useState(undefined as undefined | Error)

  useEffect(() => {
    props
      .loadSchema()
      .then((userSchema) => new Resolver().resolve(userSchema))
      .then((resolvedSchema) => mergeAllOf(resolvedSchema.result))
      .then((schemaWithoutAllOf) => setSchema(schemaWithoutAllOf))
      .catch((err) => setFetchError(err))
  }, [])

  return (
    <Layout>
      {fetchError !== undefined && (
        <div>
          <p>
            <Translate
              values={{
                errorMessage: fetchError.message,
                id: "docusaurus-json-schema-viewer-plugin.error",
              }}
            >
              {"This component crashed because of error: {errorMessage}."}
            </Translate>
          </p>
        </div>
      )}
      {schema === undefined && (
        <div>
          <p>
            <Translate values={{ id: "docusaurus-json-schema-viewer-plugin.loading" }}>
              {"Loading schema ..."}
            </Translate>
          </p>
        </div>
      )}
      {schema !== undefined && <JSONSchemaViewer schema={schema} />}
    </Layout>
  )
}

export default JSONSchemaEditor
