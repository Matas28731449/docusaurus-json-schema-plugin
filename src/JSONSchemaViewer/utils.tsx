import React from "react";

import type { JSONSchema7 } from "json-schema";

// To keep the same type everywhere here
type cleanSchema = Omit<JSONSchema7, "allOf">;

