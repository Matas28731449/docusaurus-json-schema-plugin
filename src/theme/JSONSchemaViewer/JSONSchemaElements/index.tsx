// Component to deal with oneOf / anyOf / allOf / not
export { SchemaComposition } from "./schemaComposition/index"

// Component to deal with if-then-else , dependentRequired , dependentSchemas , dependencies
export { SchemaConditional } from "./SchemaConditional/index"

// Component to deal with Object
export { CreateObject } from "./object/index"

// Component to deal with Array
export { CreateArray } from "./array/index"

// Component to deal with String
export { default as CreateString } from "./CreateString"

// Component to deal with Boolean
export { default as CreateBoolean } from "./CreateBoolean"

// Component to deal with Number
export { default as CreateNumber } from "./CreateNumber"

// Component to deal with Integer
export { default as CreateInteger } from "./CreateInteger"

// Component to deal with schema defined as "true"
export { default as CreateAlwaysValid } from "./CreateAlwaysValid"

// Component to deal with schema defined as "false"
export { default as CreateAlwaysInvalid } from "./CreateAlwaysInvalid"
