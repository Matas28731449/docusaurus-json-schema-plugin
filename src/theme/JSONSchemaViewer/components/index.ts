// To avoid circular dependancies issue, let's have a single place for exports and imports

// Collapsible componenent
export { default as Collapsible } from "./Collapsible"

// CreateDetailsNode component
export { default as CreateDetailsNode } from "./CreateDetailsNode"

// CreateEdges component
export { default as CreateEdges } from "./CreateEdges"

// CreateProperties component
export { default as CreateProperties } from "./CreateProperties"

// RenderAnyOneOf component
export { default as RenderAnyOneOf } from "./RenderAnyOneOf"

// https://stackoverflow.com/a/69328045/6149867
export type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] }
