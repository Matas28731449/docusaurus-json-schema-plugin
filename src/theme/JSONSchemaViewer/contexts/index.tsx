export * from "@theme/JSONSchemaViewer/contexts/jsvOptions"
export * from "@theme/JSONSchemaViewer/contexts/schemaHierarchy"
