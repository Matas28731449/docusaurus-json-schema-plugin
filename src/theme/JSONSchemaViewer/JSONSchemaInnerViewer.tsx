import React from "react"
import { CreateNodes } from "./components/index"

import type { JSONSchema7 } from "json-schema"

type InnerViewerProperties = {
  // Thanks to json-schema-merge-allof , we don't have allOf in the whole user schema
  // Thanks to @stoplight/json-ref-resolver, $ref are either :
  // 1. resolved
  // 2. unresolved (as circular stuff are not on the roadmap)
  schema: Omit<JSONSchema7, "allOf">
}

// External
function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  return <ul style={{ marginLeft: "1rem" }}>{CreateNodes(props.schema)}</ul>
}

export default JSONSchemaInnerViewer
