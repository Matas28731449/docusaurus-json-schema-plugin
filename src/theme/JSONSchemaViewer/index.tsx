import React, { useState, useEffect } from "react"
import { Resolver } from "@stoplight/json-ref-resolver"

import { CreateNodes, Collapsible } from "./components/index"
import { JSVOptionsContextProvider } from "./contexts/index"

import type { JSONSchema } from "./types"
import type { JSVOptions } from "./contexts/index"
import type { IResolveOpts } from "@stoplight/json-ref-resolver/types"

export type Props = {
  // The schema to use
  schema: unknown
  // To customize the ref resolving
  resolverOptions?: IResolveOpts
  // To customize the viewer
  viewerOptions?: Omit<JSVOptions, "fullSchema">
  // Other properties (who knows ?)
  [x: string]: any
}

type InnerViewerProperties = {
  // Thanks to @stoplight/json-ref-resolver, $ref are either :
  // 1. resolved
  // 2. unresolved (as circular stuff are not on the roadmap)
  schema: JSONSchema
  // Options for viewer
  viewerOptions?: Omit<JSVOptions, "fullSchema">
}

// Internal
function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  const { schema, viewerOptions } = props
  // Title of the schema, for user friendliness
  const title =
    typeof schema !== "boolean" && schema?.title !== undefined
      ? schema.title
      : "Schema"

  // state for provider
  const startingState: JSVOptions = {
    fullSchema: schema,
    // spread provided attributes
    ...viewerOptions,
  }

  return (
    <JSVOptionsContextProvider value={startingState}>
      <Collapsible
        summary={<strong>{title}</strong>}
        detailsProps={{
          open: true,
        }}
      >
        <CreateNodes schema={schema} />
      </Collapsible>
    </JSVOptionsContextProvider>
  )
}

// Entry point
export default function JSONSchemaViewer(props: Props): JSX.Element {
  const { schema: originalSchema, resolverOptions, viewerOptions } = props

  const [error, setError] = useState(undefined as undefined | Error)
  const [resolvedSchema, setResolvedSchema] = useState(
    undefined as undefined | JSONSchema
  )

  useEffect(() => {
    // Time to do the job
    new Resolver()
      .resolve(originalSchema, resolverOptions)
      .then((result) => {
        setResolvedSchema(result.result)
      })
      .catch((err) => {
        setError(err)
      })
  }, [])

  if (error !== undefined) {
    return <div>Something bad happens : {error.message}</div>
  } else if (resolvedSchema === undefined) {
    return <div>Loading ....</div>
  } else {
    return (
      <JSONSchemaInnerViewer
        schema={resolvedSchema}
        viewerOptions={viewerOptions}
      />
    )
  }
}
