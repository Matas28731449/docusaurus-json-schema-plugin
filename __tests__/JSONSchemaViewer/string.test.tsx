import React from "react"

// For typings autocomplete whatever your IDE
import { expect, test, describe } from "@jest/globals"

import { create, act } from "react-test-renderer"

import JSONSchemaViewer from "../../src/theme/JSONSchemaViewer/index"

// Type to prevent creating invalid mocks
import type { JSONSchema } from "../../src/theme/JSONSchemaViewer/types"

// Type for react-test-renderer
import type { ReactTestRenderer } from "react-test-renderer"

const testcases: [string, JSONSchema][] = [
  [
    "basic",
    {
      type: "string",
    },
  ],
  [
    "format",
    {
      type: "object",
      properties: {
        prop1: {
          type: "string",
          format: "date-time",
          examples: ["2018-11-13T20:20:39+00:00"],
        },
        prop2: {
          type: "string",
          format: "time",
          examples: ["20:20:39+00:00"],
        },
        prop3: {
          type: "string",
          format: "date",
          examples: ["2018-11-13"],
        },
        prop4: {
          type: "string",
          format: "duration",
          examples: ["P3D"],
        },
        prop5: {
          type: "string",
          format: "email",
          examples: ["test@gmail.com"],
        },
        prop6: {
          type: "string",
          format: "idn-email",
          examples: ["test@gmail.com"],
        },
        prop7: {
          type: "string",
          format: "hostname",
          examples: ["example"],
        },
        prop8: {
          type: "string",
          format: "idn-hostname",
          examples: ["example"],
        },
        prop9: {
          type: "string",
          format: "ipv4",
          examples: ["192.168.1.1"],
        },
        prop10: {
          type: "string",
          format: "ipv6",
          examples: ["2001:db8:3333:4444:5555:6666:7777:8888"],
        },
        prop11: {
          type: "string",
          format: "uuid",
          examples: ["3e4666bf-d5e5-4aa7-b8ce-cefe41c7568a"],
        },
        prop12: {
          type: "string",
          format: "uri",
          examples: ["https://www.perdu.com/"],
        },
        prop12B: {
          type: "string",
          format: "uri-reference",
        },
        prop13: {
          type: "string",
          format: "iri",
          examples: ["https://www.perdu.com/"],
        },
        prop13B: {
          type: "string",
          format: "iri-reference",
        },
        prop14: {
          type: "string",
          format: "uri-template",
        },
        prop15: {
          type: "string",
          format: "json-pointer",
        },
        prop16: {
          type: "string",
          format: "relative-json-pointer",
        },
        prop17: {
          type: "string",
          format: "regex",
          examples: ["^(\\([0-9]{3}\\))?[0-9]{3}-[0-9]{4}$"],
        },
      },
    },
  ],
  [
    "minLength",
    {
      type: "string",
      minLength: 2,
    },
  ],
  [
    "maxLength",
    {
      type: "string",
      maxLength: 3,
    },
  ],
  [
    "minLength & maxLength",
    {
      type: "string",
      minLength: 2,
      maxLength: 3,
    },
  ],
  [
    "pattern",
    {
      type: "string",
      pattern: "^(\\([0-9]{3}\\))?[0-9]{3}-[0-9]{4}$",
    },
  ],
  [
    "contentEncoding",
    {
      type: "string",
      contentEncoding: "base64",
    },
  ],
  [
    "contentMediaType",
    {
      type: "string",
      contentMediaType: "application/json",
    },
  ],
  [
    "contentEncoding with contentMediaType",
    {
      type: "string",
      contentEncoding: "base64",
      contentMediaType: "application/json",
    },
  ],
  [
    "contentSchema",
    {
      type: "string",
      contentSchema: {
        type: "object",
        required: ["name", "age"],
      },
    },
  ],
]

describe("JSONSchemaViewer - string", () => {
  test.each(testcases)("test %s", async (_title, fakeSchema) => {
    // render the component
    let root: ReactTestRenderer | undefined

    await act(async () => {
      root = create(<JSONSchemaViewer schema={fakeSchema} />)
    })

    // make assertions on root
    expect(root?.toJSON()).toMatchSnapshot()
  })
})
