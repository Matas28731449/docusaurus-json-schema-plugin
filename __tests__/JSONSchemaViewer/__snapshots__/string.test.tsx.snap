// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONSchemaViewer - string test basic 1`] = `
<details
  className="json-schema-viewer"
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        string
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - string test contentEncoding 1`] = `
<details
  className="json-schema-viewer"
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        string
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Encoding :
          </strong>
           
          <code>
            base64
          </code>
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - string test contentEncoding with contentMediaType 1`] = `
<details
  className="json-schema-viewer"
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        string
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Encoding :
          </strong>
           
          <code>
            base64
          </code>
        </div>
        <div>
          <strong>
            Media type :
          </strong>
           
          <code>
            application/json
          </code>
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - string test contentMediaType 1`] = `
<details
  className="json-schema-viewer"
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        string
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Media type :
          </strong>
           
          <code>
            application/json
          </code>
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - string test contentSchema 1`] = `
<details
  className="json-schema-viewer"
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        string
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Decoded content must be validated against this schema :
          </strong>
           
          <details
            open={true}
          >
            <summary>
              <strong>
                Schema
              </strong>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  object
                </span>
                <ul>
                  <li
                    data-level={2}
                    id="/contentSchema/properties/name"
                  >
                    <details
                      open={false}
                    >
                      <summary>
                        <strong>
                          name
                        </strong>
                         
                        <span
                          style={
                            {
                              "opacity": "0.6",
                            }
                          }
                        >
                          any
                        </span>
                         
                        <strong>
                          required
                        </strong>
                      </summary>
                      <div>
                        <div>
                          <strong>
                            type
                          </strong>
                           : 
                          <span
                            style={
                              {
                                "opacity": "0.6",
                              }
                            }
                          >
                            any
                          </span>
                          <div
                            style={
                              {
                                "marginTop": "var(--ifm-table-cell-padding)",
                              }
                            }
                          >
                            Always valid
                          </div>
                        </div>
                      </div>
                    </details>
                  </li>
                  <li
                    data-level={2}
                    id="/contentSchema/properties/age"
                  >
                    <details
                      open={false}
                    >
                      <summary>
                        <strong>
                          age
                        </strong>
                         
                        <span
                          style={
                            {
                              "opacity": "0.6",
                            }
                          }
                        >
                          any
                        </span>
                         
                        <strong>
                          required
                        </strong>
                      </summary>
                      <div>
                        <div>
                          <strong>
                            type
                          </strong>
                           : 
                          <span
                            style={
                              {
                                "opacity": "0.6",
                              }
                            }
                          >
                            any
                          </span>
                          <div
                            style={
                              {
                                "marginTop": "var(--ifm-table-cell-padding)",
                              }
                            }
                          >
                            Always valid
                          </div>
                        </div>
                      </div>
                    </details>
                  </li>
                </ul>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - string test format 1`] = `
<details
  className="json-schema-viewer"
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <ul>
        <li
          data-level={1}
          id="/properties/prop1"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop1
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                date-time
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (date-time)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop2"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop2
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                time
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (time)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop3"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop3
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                date
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (date)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop4"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop4
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                duration
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (duration)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop5"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop5
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                email
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (email)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop6"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop6
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                idn-email
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (idn-email)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop7"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop7
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                hostname
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (hostname)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop8"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop8
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                idn-hostname
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (idn-hostname)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop9"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop9
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                ipv4
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (ipv4)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop10"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop10
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                ipv6
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (ipv6)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop11"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop11
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                uuid
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (uuid)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop12"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop12
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                uri
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (uri)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop12B"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop12B
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                uri-reference
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (uri-reference)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop13"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop13
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                iri
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (iri)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop13B"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop13B
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                iri-reference
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (iri-reference)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop14"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop14
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                uri-template
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (uri-template)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop15"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop15
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                json-pointer
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (json-pointer)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop16"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop16
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                relative-json-pointer
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (relative-json-pointer)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/prop17"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                prop17
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                regex
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                   
                  (regex)
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
      </ul>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - string test maxLength 1`] = `
<details
  className="json-schema-viewer"
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        string
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Length :
          </strong>
           
          <code>
            &lt;= 3 character(s)
          </code>
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - string test minLength & maxLength 1`] = `
<details
  className="json-schema-viewer"
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        string
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Length :
          </strong>
           
          <code>
            &gt;= 2 character(s)
          </code>
           
          <strong>
            AND
          </strong>
           
          <code>
            &lt;= 3 character(s)
          </code>
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - string test minLength 1`] = `
<details
  className="json-schema-viewer"
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        string
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Length :
          </strong>
           
          <code>
            &gt;= 2 character(s)
          </code>
        </div>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - string test pattern 1`] = `
<details
  className="json-schema-viewer"
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        string
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Pattern :
          </strong>
           
          <code>
            ^(\\([0-9]{3}\\))?[0-9]{3}-[0-9]{4}$
          </code>
        </div>
      </div>
    </div>
  </div>
</details>
`;
